lib/spack/spack/cmd/diff.py

```diff
@@ -57,37 +60,36 @@ def compare_specs(a, b, to_string=False, color=None):
     Arguments:
         a (spack.spec.Spec): the first spec to compare
         b (spack.spec.Spec): the second spec to compare
-        a_name (str): the name of spec a
-        b_name (str): the name of spec b
         to_string (bool): return an object that can be json dumped
-        color (bool): whether to format the names for the console
     """
-    if color is None:
-        color = get_color_when()
 
     # Prepare a solver setup to parse differences
     setup = asp.SpackSolverSetup()
 
-    a_facts = set(t for t in setup.spec_clauses(a, body=True, expand_hashes=True))
-    b_facts = set(t for t in setup.spec_clauses(b, body=True, expand_hashes=True))
+    a = [a] if isinstance(a, spack.spec.Spec) else a
+    b = [b] if isinstance(b, spack.spec.Spec) else b
+
+    a_facts, b_facts = [], []
+    for s in a:
+        a_facts.extend(
+            setup.spec_clauses(s, body=True, transitive=True, expand_hashes=True))
+    for s in b:
+        b_facts.extend(
+            setup.spec_clauses(s, body=True, transitive=True, expand_hashes=True))
```

Comment:
I'm a little worried about a case where we lose structure in the environment because of the way the specs are conglomerated. For example, would these two environments show as equal?

```
$ cat a/spack.yaml
specs:
- tcl@8.6.11^zlib@1.2.8
- tcl@8.6.10^zlib@1.2.11

$ cat b/spack.yaml
specs:
- tcl@8.6.11^zlib@1.2.11
- tcl@8.6.10^zlib@1.2.8
```

I think those would show as identical.

But I don't think we need to resolve that for this PR, but we should document it somewhere (maybe a TODO comment) so we don't lose track of the fact there's some future work here.

Reply:
Hm, maybe I have to rethink it a bit, cause in #28885 I'm effectively doing a diff too with the same issue. A tree diff isn't terribly hard to implement as an alternative (so, a spec-constraints-diff at every level where common props are removed)

Reply:
I think part of the problem is the way we encode dependencies for clingo -- we only encode `depends_on("tcl", "zlib", "link")` or equivalent -- so there's no way to take two `tcl` and two `zlib` specs and know which is associated with which, unless you add extra metadata as you add each spec to the list.

Reply:
I need to take a closer look at this one, but I think what can get us the low hanging fruit is compare by input specs on a lockfile (at least if the final aim is #28885). For instance for the environment shown by Greg I would like to see:
```
$ spack diff ./spack.old.lock ./spack.lock 
--- /tmp/tmp.bhi1tLT0bm/spack.old.lock
+++ /tmp/tmp.bhi1tLT0bm/spack.lock
@@ input_specs @@
- tcl@8.6.11^zlib@1.2.8
- tcl@8.6.10^zlib@1.2.11
+ tcl@8.6.11^zlib@1.2.11
+ tcl@8.6.10^zlib@1.2.8
```
and then the other differences spec by spec. Would that make sense?

Reply:
That seems workable to me, @haampie ?